supervisor/process.go

```diff
@@ -0,0 +1,408 @@
+package supervisor
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/pkg/errors"
+	"os/exec"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+)
+
+const (
+	DefaultLineBufferSize = 16384
+	DefaultStreamChanSize = 200
+)
+
+type ProcessState uint8
+
+const (
+	Initial ProcessState = iota
+	Starting
+	Running
+	Stopping
+	Finished
+	Interrupted
+	Fatal
+)
+
+func (s ProcessState) String() string {
+	switch s {
+	case Initial:
+		return "initial"
+	case Starting:
+		return "starting"
+	case Running:
+		return "running"
+	case Stopping:
+		return "stopping"
+	case Finished:
+		return "finished"
+	case Interrupted:
+		return "interrupted"
+	case Fatal:
+		return "fatal"
+	default:
+		return "unknown"
+	}
+}
+
+// Process is a wrapper around exec.Cmd to run linux processes
+type Process struct {
+	*sync.Mutex
+	cmd          *exec.Cmd
+	streamStdOut chan string
+	streamStdErr chan string
+	stdOut       *StreamOutput
+	stdErr       *StreamOutput
+	state        ProcessState
+	status       Status
+	statusChan   chan Status
+	done         chan struct{}
+	stateLock    *sync.RWMutex
+}
+
+// Status is the status of a process
+type Status struct {
+	Cmd        string
+	PID        int
+	State      ProcessState
+	Error      error
+	ExitCode   int
+	StartTime  int64
+	StopTime   int64
+	StopByUser bool
+}
+
+// NewProcess creates and returns a new process without running it
+func NewProcess(name string, args []string) *Process {
+	a := strings.Join(args, " ")
+
+	status := Status{
+		Cmd:        name + " " + a,
+		PID:        0,
+		Error:      nil,
+		ExitCode:   -1,
+		StopByUser: false,
```

Comment:
nit: you can skip the zero values
```suggestion
		Cmd:        name + " " + a,
		ExitCode:   -1,
```